### ISSUE TITLE
Optimise the Delivery & Trade logic

### ISSUE THREAD

**Issue by bevans2000** [mars-sim/mars-sim#506]
**Describe the Situation**
The Trade & Delivery Mission share a large common purpose but they have different implementations in the ```TradeUtil``` & ```DeliveryUtil``` classes. This logic could be combined and improved to make it more efficient as it is computation expensive to identify potential sales activities.
The selection logic potentially has a race condition with the use of a static to hold the recently calculated best trading Settlement.
Also the logic has the potential to create ConcurrentModification exceptions as it "looks" inside the trading settlement in a non-thread safe manner from a different Settlement thread. See #484 

**Describe the Improvement**
The improvement would be to consolidate the 2 Util classes into a single logic. Coupled with this; the need to make the logic thread safe will be taken into account. This may involved creating the concept of a "Shopping  List" in the GoodManager that provides a read-only definition of a Settlement demands. This could be used by Selling Settlements and updated once a sol by the Buying Settlement.

There are probably other improvements that can be made once the work starts.

**Comment by mokun** [comment]
Btw, currently, Emergency Resupply mission is done by sending a Rover and it's too slow. It's still the only viable option for towing the damaged rover back to the settlement.

Can we send Delivery Drone instead if it's strictly missing repair parts or lacking resources ?

**Comment by mokun** [comment]
I've been having a lot of calls such as *Preloading 0 EVA suits for disembarking* non-stop:

```
00-Adir-05:079.173 (Info) RoverMission [x396] : [New Plymouth] Desert Rat - Disembarked at New Plymouth triggered by Sandra Flores.
00-Adir-05:079.173 (Info) RoverMission [x396] : [New Plymouth] Desert Rat - Preloading 0 EVA suits for disembarking.
00-Adir-05:084.907 (Info) DeliveryMeta [x2] : [Kyocera] Drone 006 available for delivery mission.
00-Adir-05:092.745 (Info) RoverMission [x264] : [Europa] Sakit - Disembarked at Europa triggered by Greta Romano.
00-Adir-05:092.745 (Info) RoverMission [x264] : [Europa] Sakit - Preloading 0 EVA suits for disembarking.
00-Adir-05:096.338 (Info) RoverMission [x177] : [New Plymouth] Desert Rat - Disembarked at New Plymouth triggered by Sandra Flores.
00-Adir-05:096.338 (Info) RoverMission [x177] : [New Plymouth] Desert Rat - Preloading 0 EVA suits for disembarking.
00-Adir-05:105.111 (Info) DeliveryMeta [x2] : [Kyocera] Drone 006 available for delivery mission.
00-Adir-05:108.272 (Info) RoverMission [x202] : [Europa] Sakit - Disembarked at Europa triggered by Greta Romano.
00-Adir-05:108.272 (Info) RoverMission [x202] : [Europa] Sakit - Preloading 0 EVA suits for disembarking

....

00-Adir-05:114.597 (Info) RoverMission [x163] : [New Plymouth] Desert Rat - Disembarked at New Plymouth triggered by Sandra Flores.
00-Adir-05:114.597 (Info) RoverMission [x163] : [New Plymouth] Desert Rat - Preloading 0 EVA suits for disembarking.
00-Adir-05:115.217 (Warning) MicroInventory [x10] : [Schiaparelli Point] Attempting to retrieve 0.014 kg water but lacking 0.004 kg.
00-Adir-05:124.178 (Info) RoverMission [x248] : [Europa] Sakit - Disembarked at Europa triggered by Greta Romano.
00-Adir-05:124.178 (Info) RoverMission [x248] : [Europa] Sakit - Preloading 0 EVA suits for disembarking.
00-Adir-05:130.923 (Info) RoverMission [x167] : [New Plymouth] Desert Rat - Disembarked at New Plymouth triggered by Sandra Flores.
00-Adir-05:130.923 (Info) RoverMission [x167] : [New Plymouth] Desert Rat - Preloading 0 EVA suits for disembarking.
```

**Comment by bevans2000** [comment]
This is due to the disembarking; not the Trade & Delivery. Nevertheless; the message can be dropped but the repeating nature is a reflection on how the disembarking logic works. There is a chunk of logic that should only be executed once when there is a transition tot eh disembarking phase. But it gets calls everytime; this includes parking a vehicle in a Garage which is why it keeps moving.

**Comment by mokun** [comment]
I'm getting messages from time to time that Gas Canister or Specimen Box could not be stored into a particular drone:
```
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 693 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) LoadingController : [Schiaparelli Point] Drone 013 - Cannot store Equipment Gas Canister 693
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 732 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 702 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 666 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 727 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 771 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 662 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 795 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 734 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 725 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 688 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 683 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 720 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 744 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 776 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 737 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 796 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 673 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 682 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 738 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 723 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 767 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 763 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 696 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 741 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 749 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 667 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 792 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 769 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 798 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 761 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 758 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 697 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 652 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 653 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 709 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 659 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 685 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Gas Canister 642 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 398 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 473 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 483 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 423 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 480 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 461 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 411 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 434 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 422 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 436 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 460 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 418 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 414 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 454 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 420 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 430 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 424 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 427 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 459 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 415 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 444 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 416 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 402 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 474 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 441 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 404 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 432 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 482 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 428 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 463 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 392 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 429 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 455 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 478 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 413 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 400 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 490 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 433 could not be stored into Drone 013.
00-Adir-02:252.720 (Warning) Equipment : Specimen Box 437 could not be stored into Drone 013.
```

**Comment by bevans2000** [comment]
It's very likely this is due to the Specimen Box being too heavy for the Drone capacity.

**Comment by mokun** [comment]
> It's very likely this is due to the Specimen Box being too heavy for the Drone capacity.

So I assume specimen box is for sale and serve as a container only ?!? 

If it's just a container, then may be we need to create a new kind of equipment that's like a shrink wrap that is featherweight in mass, instead of using Specimen Box.

**Comment by mokun** [comment]
> Secondly get rid of the singleton shared caches

which ones are the shared caches ?

**Comment by bevans2000** [comment]
Shared caches are in the Trade & Delivery classes. I'll sort these out.

**Comment by bevans2000** [comment]
An update on progress. I've done as much as I can on this and made some very localised but big changes. All the duplicate logic now resides in a simplified version in ```CommerceUtil```. I have added a new Console Command under Settlement called trade that shows all potential Trades deals for the selected Settlement. This shows that the shared logic is indeed working correctly. It also gives us an insight into what it possible.

However I've not created a Pull Request yet because there is a lack of Trade & Delivery Missions starting. But I've found out this has nothing to do with my changes but the fact that such missions never start at the beginning of the Simulation which I have noticed previously. I believe the issue is the Vehicles have a poor estimated fuel economy at the beginning and need to do a mission to get it increased. You can see this effect in the Trade command which shows the base & estimated ranges. 

I would like to create a PR but there is still a lack of Trade & Delivery missions. In the default Scenario all the Settlements are too far away to get a commerce mission going as well.

**Comment by mokun** [comment]
> the issue is the Vehicles have a poor estimated fuel economy at the beginning and need to do a mission to get it increased. You can see this effect in the Trade command which shows the base & estimated ranges.

Hmm. So in light of this, is there any unbalance you notice with regard to estimating the fuel economy ? 

If not, then we'll simply adjust those multiplying factors for boosting the probability of trade and delivery missions back to the previous level, do you agree ?

**Comment by mokun** [comment]
Btw, there's also the comparison of the value of the goods in a trade and if I remember correctly that include the value of the fuel. 

So far, the value of methane and oxygen is not high.

So I would think the fuel cost is minimal.

**Comment by mokun** [comment]
Just checked. The oxygen has a VP of 0.019. 

But methane has a VP of 43.

We could tone down the demand for methane, which is 763. It's even higher than the demand for regolith at 663.

**Comment by bevans2000** [comment]
No the problem is the estimated range for a Drone & Rover are now too low to reach any Settlement to do any Commerce. Has the calculation been changed recently that would been the estimate range has dropped  down?
In a 2 Settlement scenario I use for testing; it works because the Settlements are quite close.

**Comment by mokun** [comment]
> the estimated range for a Drone & Rover are now too low to reach any Settlement to do any Commerce. Has the calculation been changed recently that would been the estimate range has dropped down?

Yes. we need more amount methane to go per km. Also, we now need oxygen as the fuel oxidizer. 

That's probably why.

**Comment by mokun** [comment]
What's the typical km radius of a drone vs. a cargo rover right now ?

**Comment by mokun** [comment]
One possible solution is to increase the methane/oxygen tank capacity of the vehicle.

Another is to increase the drivetrain efficiency and/or fuel cell conversion efficiency in order to make the same amount of fuel travel more km per kg

**Comment by bevans2000** [comment]
Let me work through the last remaining problems then I will create a PR. The upshot is with the new trade command it is easier to see what is going on than it was before.

**Comment by mokun** [comment]
Question: will it be switching to using the price of goods ?

**Comment by bevans2000** [comment]
No if you remember we agreed to consolidate and improve the existing logic before doing that. But #601 will be much easier to do after this.

**Comment by mokun** [comment]
> consolidate and improve the existing logic before doing that. But #601 will be much easier to do after this.

Sure. 

Can you do a PR now ?

That way, it syncs up with the master branch. I've wanted to tweak GoodManager but you may or may not be working on it due to the changes in Trade and Delivery related class.

**Comment by bevans2000** [comment]
Ok. I will do it soon. Yes I've made changes to GoidManager.

**Comment by mokun** [comment]
I changed the names of some methods related to goods while streamlining and fixing the FoodInventoryTableModel yesterday. 

Hope it won't affect anything you're doing.

**Comment by bevans2000** [comment]
Should be fine. I've open a can of worm here because the new trade command is showing that sometimes the potential Trading deals are getting a profit of infinity. It's not related to these changes but I think how the Goods value have changed. The logic for calculating the Credit seems to be flawed because when calculating the new Total credit it seems to be overestimating the total credit. It over counts the number of items being transferred. 
Secondly, can you explain why in method ```GoodsManager.determineGoodValueWithSupply``` divides the value of the Good by the number of supply? That doesn't make sense because the Good.getValue is the value for a single if that Good. So for the credit we should multiple the Good.getValue by the new supply in the Settlement.

**Comment by bevans2000** [comment]
OK, I'm going to make a PR now. I've realised that I'm starting to move into #601 with the problems I'm finding. If we  simplified the whole approach and based the loading on the Cost of individual Goods as #601 says it would make more sense. The values of load is simplify:
The total of (each Good Cost * number of Goods traded)
Value is only used to build the load for trading but Cost is solely used to estimate the Credit/Profit a trade can create.
So potentially this PR will not give us more Trade/Deivery Missions but it gives us a great base to fix the way deals are defined.
Also #601 will now relate to just the new ```CommerceUtil``` class.

**Comment by mokun** [comment]
> can you explain why in method `GoodsManager.determineGoodValueWithSupply` divides the value of the Good by the number of supply? 

Do you mean why our predecessor define it as follows ? 
```
Value Point  (VP) = Demand / (1 + Supply)
```

>That doesn't make sense because the Good.getValue is the value for a single if that Good. 

To me it's just a definition thing. 

If they define VP as just a ratio of Demand-to-Supply. It should be dimension-less, while Price and Cost is [$ / kg] or [$ / item].

> So for the credit we should multiple the Good.getValue by the new supply in the Settlement.

But the question is, what does the `new VP = Demand * Supply` would serve ?

If a settlement demands 10 apples but at the same time can produce 10 apples, then the ratio is 1 and it makes sense.

But knowing just `10 * 10 = 100` won't do much, right ?

**Comment by bevans2000** [comment]
Moving gthis discussion to #601

**Comment by mokun** [comment]
There are NPE as follows: 
```
00-Adir-06:851.137 (Info) CommerceUtil : [New Plymouth] New best deal for DELIVERY to Schiaparelli Point profit 41314.65196778025
00-Adir-06:851.137 (Info) DeliveryMeta : [New Plymouth] Joshua Murphy - DeliveryMeta's probability: 400.0
00-Adir-06:851.137 (Info) MissionManager : [New Plymouth] Joshua Murphy - Mission 'Delivery' probability=400.0 base prob=400.0 boost=0.0 sponsor=1.0
00-Adir-06:851.137 (Info) MissionManager : [New Plymouth] Joshua Murphy - Mission 'Emergency Supply' probability=100.0 base prob=100.0 boost=0.0 sponsor=1.0
00-Adir-06:851.137 (Severe) PersonTaskManager : [New Plymouth] Joshua Murphy - Trouble calling performTask(): 
java.lang.NullPointerException
	at org.mars_sim.msp.core.goods.CommerceUtil.getPotentialDeal(CommerceUtil.java:122)
	at org.mars_sim.msp.core.goods.CommerceUtil.getBestDeal(CommerceUtil.java:93)
	at org.mars_sim.msp.core.goods.GoodsManager.getBestDeal(GoodsManager.java:741)
	at org.mars_sim.msp.core.person.ai.mission.meta.TradeMeta.getSettlementProbability(TradeMeta.java:102)
	at org.mars_sim.msp.core.person.ai.mission.meta.TradeMeta.getProbability(TradeMeta.java:70)
	at org.mars_sim.msp.core.person.ai.mission.MissionManager.getNewMission(MissionManager.java:268)
	at org.mars_sim.msp.core.person.ai.Mind.getNewMission(Mind.java:479)
	at org.mars_sim.msp.core.person.ai.task.PlanMission.selectingPhase(PlanMission.java:134)
	at org.mars_sim.msp.core.person.ai.task.PlanMission.performMappedPhase(PlanMission.java:111)
	at org.mars_sim.msp.core.person.ai.task.utils.Task.executeMappedPhase(Task.java:629)
	at org.mars_sim.msp.core.person.ai.task.utils.Task.performTask(Task.java:584)
	at org.mars_sim.msp.core.person.ai.task.utils.PersonTaskManager.executeTask(PersonTaskManager.java:101)
	at org.mars_sim.msp.core.person.ai.Mind.takeAction(Mind.java:185)
	at org.mars_sim.msp.core.person.ai.Mind.timePassing(Mind.java:123)
	at org.mars_sim.msp.core.person.Person.timePassing(Person.java:841)
	at org.mars_sim.msp.core.structure.Settlement.timePassing(Settlement.java:1114)
	at org.mars_sim.msp.core.structure.Settlement.otherTimePassings(Settlement.java:998)
	at org.mars_sim.msp.core.structure.Settlement.timePassing(Settlement.java:953)
	at org.mars_sim.msp.core.UnitManager$SettlementTask.call(UnitManager.java:753)
	at org.mars_sim.msp.core.UnitManager$SettlementTask.call(UnitManager.java:1)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Comment by bevans2000** [comment]
I know what this is. The Settlement has no available Vehicles; probably only occurs on longer simulations.

**Comment by mokun** [comment]
Another NPE: 
```

00-Adir-02:044.481 (Info) DeliveryMeta : [Schiaparelli Point] Drone 017 - Available for delivery mission.
00-Adir-02:044.481 (Severe) PersonTaskManager : [Schiaparelli Point] Ryuto Yamashita - Trouble calling performTask(): 
java.lang.NullPointerException
	at org.mars_sim.msp.core.goods.GoodsManager.getBestDeal(GoodsManager.java:733)
	at org.mars_sim.msp.core.person.ai.mission.meta.DeliveryMeta.getSettlementProbability(DeliveryMeta.java:113)
	at org.mars_sim.msp.core.person.ai.mission.meta.DeliveryMeta.getProbability(DeliveryMeta.java:68)
	at org.mars_sim.msp.core.person.ai.mission.MissionManager.getNewMission(MissionManager.java:268)
	at org.mars_sim.msp.core.person.ai.Mind.getNewMission(Mind.java:479)
	at org.mars_sim.msp.core.person.ai.task.PlanMission.selectingPhase(PlanMission.java:134)
	at org.mars_sim.msp.core.person.ai.task.PlanMission.performMappedPhase(PlanMission.java:111)
	at org.mars_sim.msp.core.person.ai.task.utils.Task.executeMappedPhase(Task.java:629)
	at org.mars_sim.msp.core.person.ai.task.utils.Task.performTask(Task.java:584)
	at org.mars_sim.msp.core.person.ai.task.utils.PersonTaskManager.executeTask(PersonTaskManager.java:101)
	at org.mars_sim.msp.core.person.ai.Mind.takeAction(Mind.java:185)
	at org.mars_sim.msp.core.person.ai.Mind.timePassing(Mind.java:123)
	at org.mars_sim.msp.core.person.Person.timePassing(Person.java:841)
	at org.mars_sim.msp.core.structure.Settlement.timePassing(Settlement.java:1114)
	at org.mars_sim.msp.core.structure.Settlement.otherTimePassings(Settlement.java:998)
	at org.mars_sim.msp.core.structure.Settlement.timePassing(Settlement.java:953)
	at org.mars_sim.msp.core.UnitManager$SettlementTask.call(UnitManager.java:753)
	at org.mars_sim.msp.core.UnitManager$SettlementTask.call(UnitManager.java:1)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Comment by bevans2000** [comment]
Odd, I'm not seeing any of these. I said yesterday I thought it wasn't ready for a PR. Should of kept it on the branch for more testing.

**Comment by bevans2000** [comment]
You've saved and reloaded; that's why.

**Comment by mokun** [comment]
> Odd, I'm not seeing any of these. I said yesterday I thought it wasn't ready for a PR. Should of kept it on the branch for more testing.

Not a problem.

**Comment by mokun** [comment]
I have this: 

```
00-Adir-06:004.176 (Severe) PersonTaskManager : [Tian Cheng] Peijing Deng - Trouble calling performTask(): 
java.util.ConcurrentModificationException
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1626)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.ReduceOps$5.evaluateSequential(ReduceOps.java:257)
	at java.base/java.util.stream.ReduceOps$5.evaluateSequential(ReduceOps.java:248)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.count(ReferencePipeline.java:605)
	at org.mars_sim.msp.core.equipment.EquipmentInventory.findNumEmptyContainersOfType(EquipmentInventory.java:340)
	at org.mars_sim.msp.core.structure.Settlement.findNumEmptyContainersOfType(Settlement.java:3846)
	at org.mars_sim.msp.core.goods.CommerceUtil.getNumInInventory(CommerceUtil.java:661)
	at org.mars_sim.msp.core.goods.CommerceUtil.getGoodValue(CommerceUtil.java:550)
	at org.mars_sim.msp.core.goods.CommerceUtil.findBestGood(CommerceUtil.java:451)
	at org.mars_sim.msp.core.goods.CommerceUtil.determineLoad(CommerceUtil.java:310)
	at org.mars_sim.msp.core.goods.CommerceUtil.getDesiredBuyLoad(CommerceUtil.java:234)
	at org.mars_sim.msp.core.person.ai.mission.Delivery.<init>(Delivery.java:107)
	at org.mars_sim.msp.core.person.ai.mission.meta.DeliveryMeta.constructInstance(DeliveryMeta.java:42)
	at org.mars_sim.msp.core.person.ai.mission.MissionManager.getNewMission(MissionManager.java:320)
	at org.mars_sim.msp.core.person.ai.Mind.getNewMission(Mind.java:479)
	at org.mars_sim.msp.core.person.ai.task.PlanMission.selectingPhase(PlanMission.java:134)
	at org.mars_sim.msp.core.person.ai.task.PlanMission.performMappedPhase(PlanMission.java:111)
	at org.mars_sim.msp.core.person.ai.task.utils.Task.executeMappedPhase(Task.java:629)
	at org.mars_sim.msp.core.person.ai.task.utils.Task.performTask(Task.java:584)
	at org.mars_sim.msp.core.person.ai.task.utils.PersonTaskManager.executeTask(PersonTaskManager.java:101)
	at org.mars_sim.msp.core.person.ai.Mind.takeAction(Mind.java:185)
	at org.mars_sim.msp.core.person.ai.Mind.timePassing(Mind.java:123)
	at org.mars_sim.msp.core.person.Person.timePassing(Person.java:841)
	at org.mars_sim.msp.core.structure.Settlement.timePassing(Settlement.java:1112)
	at org.mars_sim.msp.core.structure.Settlement.otherTimePassings(Settlement.java:996)
	at org.mars_sim.msp.core.structure.Settlement.timePassing(Settlement.java:951)
	at org.mars_sim.msp.core.UnitManager$SettlementTask.call(UnitManager.java:753)
	at org.mars_sim.msp.core.UnitManager$SettlementTask.call(UnitManager.java:1)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Comment by bevans2000** [comment]
Ok, I've got some changes pending for this

**Comment by bevans2000** [comment]
@mokun How many Settlements were running?

**Comment by mokun** [comment]
> @mokun How many Settlements were running?

I have all 9

**Comment by bevans2000** [comment]
Right, so this is probably the same problem we've seen before where there is a Delivery/Trade mission doing the trade at the remote site in one thread whilst the "home" thread is also updating the inventory. I've got a change that reduces how often the trade load is recalculated.
But essentially to fully solve it we need to implement the idea I proposed earlier in the other thread. The GoodManager creates a "shopping list" of needed ideas that selling Settlements use instead of them each recalculating the scores on demand; hence increasing the chance of concurrency problem.

**Comment by mokun** [comment]
> GoodManager creates a "shopping list" of needed ideas that selling Settlements use instead of them each recalculating the scores on demand; hence increasing the chance of concurrency problem.

So what I don't understand is in case like this, how you would completely remove concurrency issue ? 

Say settlement A is on thread 1 and settlement B is on thread 2.

I suppose our goal is to avoid the temptation of having settlement A accessing thread 2.

Similarly, we want to avoid having settlement B accessing thread 1.

There is two GoodManagers now. One on settlement A, one on settlement B.

**Comment by bevans2000** [comment]
Yes so the "home" thread #1 is the only one that can change anything in Settlement A and the same for Thread #2 & Settlement B. But Thread #1 could read something in Settlement B as long as it was stable and not changing. This is the general principle for any Concurrent modification problem.
So the shopping list in Settlement A is only ever updated by Thread #1 probably periodically, e.g. twice a Sol. It creates a new shoppinglist in parallel that only it can see. Then once fully completed and stable it "publishes".
In this model, the Remote Thread #2 would only ever read an stable published shopping list. Even if it was changing during the read, Thread #2 would be using the original version.
But what's a shopping list? I think it is an ordered list of Goods that the remote settlement needs. Then potential sellers just walk through the fixed static list.

**Comment by bevans2000** [comment]
I was trying to avoid doing this but now I realise that we can't avoid it because the current approach has this concurrent design flaw which gets worse when there are more Settlement in action. It's not actually that much work so I will make a start tonight.

**Comment by mokun** [comment]
It seems to me that in the old approach, Thread 1 is modifying the mutable data on behalf of settlement A and they are stored on its own version of Delivery or Trade and its Settlement A object

The same is true with Thread 2 on its own version of Delivery/Trade and its Settlement B object

But I suppose in the existing approach, thread 1 would try to change things on Settlement B while thread 2 would change things on Settlement A and that's where the problems lie ?

**Comment by bevans2000** [comment]
Not exactly. The scenarios is Thread 1 is creating a Delivery for Settlement A so it indirectly scans the inventory of Settlement B to calculate the buy load. In parallel Thread 2 updates the Inventory of Settlement B, e.g. a Vehicle is unloaded, and then changes the inventory and hence creates the concurrent modification issue. Thread 2 should be able to update anything in Settlement B as it's the home thread.

**Comment by mokun** [comment]
So where do you put that stable published shopping list both thread 1 and thread 2 to access (read and write) safely ?

**Comment by mokun** [comment]
I supposed there's a way to create a set of immutable data (on a xml/json file, etc.). 

Once it's created, it can't be changed. So it's thread-safe with no concurrent issue.

This way, you can even read it later using console or Swing UI or open it with any xml reader.

**Comment by bevans2000** [comment]
I've had a look and to fix this correctly the problem is bigger because the root cause is how trading is calculated with a Selling Settlement being allowed to look inside a Buying settlement. In a business world, this never happens, the buyer says what they want and a seller decided whether to give it to them. Essentially we treat the seller & buyer the same in out Commerce calculations which is us to read "private" attribute in the buying Settlement.
My thoughts are:

1. We should change getBuyList in Goodsmanager. This should be an instance method and return what that Settlement needs not the static list of tradable Goods.
2. Each item in the Buy list would have to be a combination of a Good, how many the Settlement wants & how much they want to pay
3. When deciding a Load the buyer's buy list would be compared to the local seller's Inventory. If the seller has enough of a certain Good then they sell it.  This might compare the local Price to the Buyers prices to decide how many to sell, if any.
4. The GoodsManager Buylist would have to be updated periodically; maybe twice per sol. We could use Good.getDemandValue to scale how much is needed. 
5. The buying price we could pro-rated an increase on the Good base price using the demand.
6. Ideally the price/Cost of a Good should be different at each Settlement but that's a bugger activity. This approach would takes us down that route of true Supply/Demand pricing.

@mokun What do you think? I like the approach but there could be some tuning needed.
The BuyList would be combined to the existing Good tab in the Monitor Tool.
This is overlapping with #601

**Comment by mokun** [comment]
> 1. We should change getBuyList in Goodsmanager. This should be an instance method and return what that Settlement needs not the static list of tradable Goods.

Sure. How about we have Settlement A comes up with a list of goods together with the maximum price on each that it would accept. 

The maximum price could be something like 20% above the price of the settlement A OR above the ***Inter-market** price of each good.

We could also allow this max price to be player customizable in Swing UI. 

Settlement B won't know about this max price. It's hidden from settlement B. 

During negotiation, a high trading skill trader in settlement B would be able to get the selling price as close to that max price as possible.

On the other hand, a high trading skill trader in settlement A should be able to get a buying price that's as low as possible.

**Comment by mokun** [comment]
We could allow player to use the CommanderWindow to make changes. 

btw, *Command Mode* is not working right now. 

I'd say *Command Mode* is a good idea to have a player focus on only one settlement at a time 

Or we can even create *Sponsor Mode* in that a player will play a sponsor that will govern over a few settlements. 

However, we are by far only focusing a lot of time in making the *Sandbox Mode* stable.


we may as well

**Comment by bevans2000** [comment]
I like the idea of reducing the price with a high skilled negotiator.
Thinking of #601 it makes sense to change the approach. I'm going to create a branch for these changes to give time to refine them.
Also if the buyer is the same Reporting Authority there may be no cost to the deal?

**Comment by bevans2000** [comment]
Plodding on with this big activity slowly due to lack of time :-(

Simplified and optimise the approach to a better model that is easier to understand. The principle now revolves around each  Settlement having a list of Goods it wants to buy and a list of Goods it wants to sell; each have a quantity and price. Then a Deal is simplify the intersection of the 2 lists where the Buy price is higher than the Sell price, simple. Avoids concurrency issue because the home Settlement maintains its own lists.
The tricky part that I haven't totally bottomed out is how to decide what to sell and what to buy.

**Comment by mokun** [comment]
> The tricky part that I haven't totally bottomed out is how to decide what to sell and what to buy.

We have indeed been trying to automate too many processes in mars-sim. 

So may be we can proceed slowly in redesigning the negotiation because we don't want to hide those  assumptions anymore as in the past.

We may as well show all the modifiers by designing an UI window to show what's going on behind the scene as much as possible. 

Say we create a modal window to pop up during the negotiation to bring that up to player's attention.

This modal window will pause the simulation if player agreed with its default setting in Commander Dashboard.

**Comment by bevans2000** [comment]
Maybe but I do feel that it should be automated with the user influencing the approach but not being mandatory.
Anyway I've nearly reached the end of this now. It's a lot better in terms of running and clarity but needs more tuning.
